cmake/CompileBoost.cmake

```diff
@@ -1,26 +1,97 @@
-find_package(Boost 1.72)
+function(compile_boost)
+  set(options)
+  set(oneValueArgs TARGET)
+  set(multiValueArgs BUILD_ARGS CXXFLAGS LDFLAGS)
+  cmake_parse_arguments(MY "${options}" "${oneValueArgs}"
+                          "${multiValueArgs}" ${ARGN} )
+  # Configure the boost toolset to use
+  set(BOOTSTRAP_COMMAND "./bootstrap.sh --with-libraries=context")
+  set(B2_COMMAND "./b2")
+  set(BOOST_COMPILER_FLAGS -fvisibility=hidden -fPIC -std=c++14 -w)
+  set(BOOST_CXX_COMPILER "${CMAKE_CXX_COMPILER}")
+  if(APPLE)
+    set(BOOST_TOOLSET "darwin")
+    # this is to fix a weird macOS issue -- by default
+    # cmake would otherwise pass a compiler that can't
+    # compile boost
+    set(BOOST_CXX_COMPILER "/usr/bin/clang++")
+  elseif(CLANG)
+    set(BOOST_TOOLSET "clang")
+    set(BOOTSTRAP_COMMAND "${BOOTSTRAP_COMMAND} --with-toolset=clang")
+  else()
+    set(BOOST_TOOLSET "gcc")
+  endif()
+  if(APPLE OR USE_LIBCXX)
+    list(APPEND BOOST_COMPILER_FLAGS -stdlib=libc++)
+  endif()
+  set(BOOST_ADDITIONAL_COMPILE_OPTIOINS "")
+  foreach(flag IN LISTS BOOST_COMPILER_FLAGS MY_CXXFLAGS)
+    string(APPEND BOOST_ADDITIONAL_COMPILE_OPTIOINS "<cxxflags>${flag} ")
+  endforeach()
+  foreach(flag IN LISTS MY_LDFLAGS)
+    string(APPEND BOOST_ADDITIONAL_COMPILE_OPTIOINS "<linkflags>${flag} ")
+  endforeach()
+  configure_file(${CMAKE_SOURCE_DIR}/cmake/user-config.jam.cmake ${CMAKE_BINARY_DIR}/user-config.jam)
+
+  set(USER_CONFIG_FLAG --user-config=${CMAKE_BINARY_DIR}/user-config.jam)
 
-if(Boost_FOUND)
-  add_library(boost_target INTERFACE)
-  target_link_libraries(boost_target INTERFACE Boost::boost)
-else()
   include(ExternalProject)
-  ExternalProject_add(boostProject
+  set(BOOST_INSTALL_DIR "${CMAKE_BINARY_DIR}/boost_install")
+  ExternalProject_add("${MY_TARGET}Project"
     URL "https://dl.bintray.com/boostorg/release/1.72.0/source/boost_1_72_0.tar.bz2"
     URL_HASH SHA256=59c9b274bc451cf91a9ba1dd2c7fdcaf5d60b1b3aa83f2c9fa143417cc660722
-    CONFIGURE_COMMAND ""
-    BUILD_COMMAND ""
+    CONFIGURE_COMMAND ${BOOTSTRAP_COMMAND} --with-libraries=context
```

Comment:
Based on the CI error messages it looks like the entire string ${BOOTSTRAP_COMMAND} is being considered an argument, rather than getting tokenized. It appears that `CONFIGURE_COMMAND` accepts a `;`-delimited cmake-style list, although I don't see that explicitly mentioned in the cmake documentation.

Reply:
ah yes this makes sense (I don't think this needs to be mentioned in the docs -- this is just how this works in cmake...).

This whole section got a bit ugly because at one point I naively thought I could make this work on Windows (which of course, I didn't -- so on Windows the user has to install this first). I'll fix this and make it a bit nicer.